tools/fm-editor/graph/FeatureModelGraph.cpp

```diff
@@ -0,0 +1,217 @@
+#include "FeatureModelGraph.h"
+#include "vara/Feature/Feature.h"
+#include "vara/Feature/FeatureModel.h"
+#include "vara/Feature/FeatureModelTransaction.h"
+
+#include <QKeyEvent>
+#include <QRandomGenerator>
+
+#include <cmath>
+
+using vara::feature::Feature;
+
+FeatureModelGraph::FeatureModelGraph(vara::feature::FeatureModel *FeatureModel,
+                                     QWidget *Parent)
+    : QGraphicsView(Parent),
+      EntryNode(new FeatureNode(FeatureModel->getRoot())),
+      FeatureModel(FeatureModel) {
+
+  Scene = std::make_unique<QGraphicsScene>(this);
+  Scene->setItemIndexMethod(QGraphicsScene::NoIndex);
+  setScene(Scene.get());
+  setCacheMode(CacheBackground);
+  setViewportUpdateMode(BoundingRectViewportUpdate);
+  setRenderHint(QPainter::Antialiasing);
+  setTransformationAnchor(AnchorUnderMouse);
+  scale(qreal(0.8), qreal(0.8));
+  reload();
+  setDragMode(ScrollHandDrag);
+  Scene->setSceneRect(0, 0, EntryNode->childrenWidth() + 100,
+                      100 * EntryNode->childrenDepth() + 100);
+}
+
+void FeatureModelGraph::reload() {
+  Nodes.push_back(std::unique_ptr<FeatureNode>(EntryNode));
+  auto *Scene = this->scene();
+  Scene->clear();
+  Scene->addItem(EntryNode);
+  buildRec(EntryNode);
+  auto NextChildren = std::vector<FeatureNode *>(EntryNode->children().size());
+  auto CurrentChildren = EntryNode->children();
+  std::transform(CurrentChildren.begin(), CurrentChildren.end(),
+                 NextChildren.begin(),
+                 [](FeatureEdge *Edge) { return Edge->targetNode(); });
+  positionRec(1, NextChildren, 5);
+  EntryNode->setPos(EntryNode->childrenWidth() / 2.0, 10);
+}
+
+void FeatureModelGraph::buildRec(FeatureNode *CurrentFeatureNode) {
+  for (auto *Feature :
+       CurrentFeatureNode->getFeature()->getChildren<vara::feature::Feature>(
+           1)) {
+    auto Node = std::make_unique<FeatureNode>(Feature);
+    auto *Edge = new FeatureEdge(CurrentFeatureNode, Node.get());
+    scene()->addItem(Edge);
+    scene()->addItem(Node.get());
```

Comment:
The destructor of `QGraphicsScene` will call `delete` on all the items that were added to the scene.

From what I see this is probably the cause for the segfaults when loading a new feature model, as when reassigning the `FeatureModelGraph` (FeatureModelEditor.cpp, `buildGraph`), the `FeatureModelGraph::Nodes` gets destroyed, however the `FeatureNode`s inside it have already been deleted by the deconstructor of the `QGraphicsScene`